selftest/test_images.py

```diff
@@ -7,129 +7,428 @@
-
-
-def subprocess_run(cmd: List[str], check: bool = True):
+if not GEN2_VM_SIZES:
+    GEN2_VM_SIZES = [
+        sku.vm_size
+        for sku in selftest.SKU_CONFIGS.values()
+        if sku.vm_size_type == "x64"
+    ]
+
+
+def safe_parse_bytes_literal(s: str):
+    """Parse a bytes literal string (e.g. b'foo') safely to a string (foo)."""
+    s = s.strip()
+    if not s.startswith("b'") or not s.endswith("'"):
+        return s
+
+    # Strip the b' and trailing '
+    inner = s[2:-1]
+    # Decode escape sequences (e.g., \r, \n, \x00)
+    raw_bytes: bytes = codecs.escape_decode(inner.encode())[0]  # type: ignore
+    return raw_bytes.decode("utf-8", errors="ignore").rstrip("\x00")
+
+
+def _subprocess_run(
+    cmd: List[str],
+    *,
+    artifacts_path: Path,
+    artifact_name: str,
+    check: bool = False,
+    decode: bool = False,
+) -> subprocess.CompletedProcess:
     """Run a subprocess command and capture outputs as utf-8."""
-    print(f"executing command: {shlex.join(cmd)}")
-    try:
+    artifact_path = artifacts_path / artifact_name
+    deadline = datetime.now(timezone.utc) + timedelta(minutes=30)
+    while datetime.now(timezone.utc) < deadline:
+        printable_cmd = shlex.join(cmd)
+        logger.debug("executing command: %s", printable_cmd)
         proc = subprocess.run(
             cmd,
-            check=check,
+            check=False,
             capture_output=True,
             text=True,
             encoding="utf-8",
+            errors="replace",
+        )
+        logger.debug(
+            "executed command=%s rc=%d artifact=%s",
+            printable_cmd,
+            proc.returncode,
+            artifact_path.as_posix(),
         )
-    except subprocess.CalledProcessError as error:
-        print(
-            f"error running command: {error} stdout={error.stdout} stderr={error.stderr}"
+
+        stdout = proc.stdout
+        stderr = proc.stderr
+        if decode:
+            # Decode potential byte repr strings from get-boot-log.
+            stdout = safe_parse_bytes_literal(stdout)
+            stderr = safe_parse_bytes_literal(stderr)
+
+        artifact_path.write_text(
+            f"cmd: {printable_cmd}\nrc: {proc.returncode}\nstdout:\n{stdout!s}\nstderr:\n{stderr!s}",
+            encoding="utf-8",
         )
-        raise
 
-    print(f"executed command: {proc}")
-    return proc
+        if proc.returncode != 0 and proc.stderr.startswith("ssh:"):
+            logger.debug("SSH failed, retrying...")
+            time.sleep(5)
+            continue
+
+        if check and proc.returncode != 0:
+            raise RuntimeError(
+                f"Command '{printable_cmd}' failed with return code {proc.returncode}: {artifact_path.as_posix()}"
+            )
+
+        return proc
+
+    raise RuntimeError(
+        f"Command '{shlex.join(cmd)}' timed out after 30 minutes: {artifacts_path.as_posix()}"
```

Comment:
After hoising `printable_cmd` outside loop
```suggestion
        f"Command '{printable_cmd}' timed out after 30 minutes: {artifacts_path.as_posix()}"
```

Reply:
fixed!